File: sdk/typescript/src/exporters/ocsf-exporter.ts
/**
 * AITF OCSF Exporter.
 *
 * OTel SpanExporter that converts AI spans to OCSF Category 7 events
 * and exports them to SIEM/XDR endpoints, S3, or local files.
 *
 * Based on forwarder architecture from the AITelemetry project.
 */

import * as fs from "fs";
import * as path from "path";
import * as http from "http";
import * as https from "https";
import { URL } from "url";
import { ReadableSpan, SpanExporter } from "@opentelemetry/sdk-trace-base";
import { ExportResult, ExportResultCode } from "@opentelemetry/core";
import { OCSFMapper } from "../ocsf/mapper";
import { ComplianceMapper } from "../ocsf/compliance-mapper";
import { AIBaseEvent, stripNulls } from "../ocsf/schema";

/** Options for configuring the OCSFExporter. */
export interface OCSFExporterOptions {
  endpoint?: string;
  outputFile?: string;
  complianceFrameworks?: string[];
  includeRawSpan?: boolean;
  apiKey?: string;
}

/**
 * Exports OTel spans as OCSF Category 7 AI events.
 *
 * Converts AI-related spans to OCSF events using OCSFMapper,
 * enriches with compliance metadata, and exports to configured
 * destinations.
 *
 * Usage:
 *   import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
 *   import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
 *   import { OCSFExporter } from "@aitf/sdk";
 *
 *   const exporter = new OCSFExporter({
 *     outputFile: "/var/log/aitf/ocsf_events.jsonl",
 *     complianceFrameworks: ["nist_ai_rmf", "eu_ai_act", "mitre_atlas"],
 *   });
 *   const provider = new NodeTracerProvider();
 *   provider.addSpanProcessor(new BatchSpanProcessor(exporter));
 */
export class OCSFExporter implements SpanExporter {
  private readonly _endpoint: string | null;
  private readonly _outputFile: string | null;
  private readonly _includeRawSpan: boolean;
  private readonly _apiKey: string | null;
  private readonly _mapper: OCSFMapper;
  private readonly _complianceMapper: ComplianceMapper;
  private _eventCount = 0;

  constructor(options: OCSFExporterOptions = {}) {
    this._endpoint = options.endpoint ?? null;
    this._outputFile = options.outputFile ?? null;
    this._includeRawSpan = options.includeRawSpan ?? false;
    this._apiKey = options.apiKey ?? null;
    this._mapper = new OCSFMapper();
    this._complianceMapper = new ComplianceMapper({
      frameworks: options.complianceFrameworks,
    });

    // Ensure output directory exists
    if (this._outputFile) {
      const dir = path.dirname(this._outputFile);
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  /**
   * Export spans as OCSF events.
   */
  export(
    spans: ReadableSpan[],
    resultCallback: (result: ExportResult) => void
  ): void {
    const events: Record<string, unknown>[] = [];

    for (const span of spans) {
      const ocsfEvent = this._mapper.mapSpan(span);
      if (!ocsfEvent) {
        continue;
      }

      // Enrich with compliance metadata
      const eventType = this._classifyEvent(ocsfEvent);
      if (eventType) {
        this._complianceMapper.enrichEvent(ocsfEvent, eventType);
      }

      const eventDict = stripNulls(
        ocsfEvent as unknown as Record<string, unknown>
      );
      events.push(eventDict);
      this._eventCount++;
    }

    if (events.length === 0) {
      resultCallback({ code: ExportResultCode.SUCCESS });
      return;
    }

    try {
      if (this._outputFile) {
        this._exportToFile(events);
      }
      if (this._endpoint) {
        this._exportToEndpoint(events)
          .then(() => {
            resultCallback({ code: ExportResultCode.SUCCESS });
          })
          .catch((err) => {
            console.error("OCSF export to endpoint failed:", err);
            resultCallback({ code: ExportResultCode.FAILED });
          });
        return;
      }
      resultCallback({ code: ExportResultCode.SUCCESS });
    } catch (err) {
      console.error("OCSF export failed:", err);
      resultCallback({ code: ExportResultCode.FAILED });
    }
  }

  private _exportToFile(events: Record<string, unknown>[]): void {
    if (!this._outputFile) return;

    const lines = events
      .map((event) => JSON.stringify(event))
      .join("\n");
    fs.appendFileSync(this._outputFile, lines + "\n", "utf-8");
  }

  private async _exportToEndpoint(
    events: Record<string, unknown>[]
  ): Promise<void> {
    if (!this._endpoint) return;

    const url = new URL(this._endpoint);
    const payload = JSON.stringify(events);
    const isHttps = url.protocol === "https:";

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "Content-Length": String(Buffer.byteLength(payload)),
    };
    if (this._apiKey) {
      headers["Authorization"] = `Bearer ${this._apiKey}`;
    }

    return new Promise<void>((resolve, reject) => {
      const requestOptions: http.RequestOptions = {
        hostname: url.hostname,
        port: url.port,
        path: url.pathname + url.search,
        method: "POST",
        headers,
        timeout: 30000,
      };

      const transport = isHttps ? https : http;
      const req = transport.request(requestOptions, (res) => {
        if (res.statusCode && res.statusCode >= 400) {
          reject(
            new Error(
              `OCSF endpoint returned ${res.statusCode}`
            )
          );
        } else {
          resolve();
        }
        // Consume response data to free memory
        res.resume();
      });

      req.on("error", reject);
      req.on("timeout", () => {
        req.destroy();
        reject(new Error("OCSF endpoint request timed out"));
      });

      req.write(payload);
      req.end();
    });
  }

  private _classifyEvent(event: AIBaseEvent): string | null {
    const classUid = event.class_uid;
    const mapping: Record<number, string> = {
      7001: "model_inference",
      7002: "agent_activity",
      7003: "tool_execution",
      7004: "data_retrieval",
      7005: "security_finding",
      7006: "supply_chain",
      7007: "governance",
      7008: "identity",
    };
    return mapping[classUid] ?? null;
  }

  /** Number of OCSF events exported. */
  get eventCount(): number {
    return this._eventCount;
  }

  async shutdown(): Promise<void> {
    // No-op
  }

  async forceFlush(): Promise<void> {
    // No-op
  }
}

File: sdk/python/src/aitf/exporters/ocsf_exporter.py
"""AITF OCSF Exporter.

OTel SpanExporter that converts AI spans to OCSF Category 7 events
and exports them to SIEM/XDR endpoints, S3, or local files.

Based on forwarder architecture from the AITelemetry project.
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Sequence

from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult

from aitf.ocsf.compliance_mapper import ComplianceMapper
from aitf.ocsf.mapper import OCSFMapper
from aitf.ocsf.schema import AIBaseEvent

logger = logging.getLogger(__name__)


class OCSFExporter(SpanExporter):
    """Exports OTel spans as OCSF Category 7 AI events.

    Converts AI-related spans to OCSF events using OCSFMapper,
    enriches with compliance metadata, and exports to configured
    destinations.

    Usage:
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        exporter = OCSFExporter(
            output_file="/var/log/aitf/ocsf_events.jsonl",
            compliance_frameworks=["nist_ai_rmf", "eu_ai_act", "mitre_atlas"],
        )
        provider = TracerProvider()
        provider.add_span_processor(BatchSpanProcessor(exporter))
    """

    def __init__(
        self,
        endpoint: str | None = None,
        output_file: str | None = None,
        compliance_frameworks: list[str] | None = None,
        include_raw_span: bool = False,
        api_key: str | None = None,
    ):
        self._endpoint = endpoint
        self._output_file = output_file
        self._include_raw_span = include_raw_span
        self._api_key = api_key
        self._mapper = OCSFMapper()
        self._compliance_mapper = ComplianceMapper(frameworks=compliance_frameworks)
        self._event_count = 0

        if output_file:
            Path(output_file).parent.mkdir(parents=True, exist_ok=True)

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        """Export spans as OCSF events."""
        events: list[dict[str, Any]] = []

        for span in spans:
            ocsf_event = self._mapper.map_span(span)
            if ocsf_event is None:
                continue

            # Enrich with compliance metadata
            event_type = self._classify_event(ocsf_event)
            if event_type:
                self._compliance_mapper.enrich_event(ocsf_event, event_type)

            event_dict = ocsf_event.model_dump(exclude_none=True)
            events.append(event_dict)
            self._event_count += 1

        if not events:
            return SpanExportResult.SUCCESS

        # Export to configured destinations
        try:
            if self._output_file:
                self._export_to_file(events)
            if self._endpoint:
                self._export_to_endpoint(events)
            return SpanExportResult.SUCCESS
        except Exception as exc:
            logger.error("OCSF export failed: %s", exc)
            return SpanExportResult.FAILURE

    def _export_to_file(self, events: list[dict[str, Any]]) -> None:
        """Write OCSF events to JSONL file."""
        with open(self._output_file, "a") as f:
            for event in events:
                f.write(json.dumps(event, default=str) + "\n")

    def _export_to_endpoint(self, events: list[dict[str, Any]]) -> None:
        """Send OCSF events to HTTP endpoint."""
        try:
            import urllib.request

            headers = {"Content-Type": "application/json"}
            if self._api_key:
                headers["Authorization"] = f"Bearer {self._api_key}"

            payload = json.dumps(events, default=str).encode("utf-8")
            req = urllib.request.Request(
                self._endpoint,
                data=payload,
                headers=headers,
                method="POST",
            )
            with urllib.request.urlopen(req, timeout=30) as resp:
                if resp.status >= 400:
                    logger.error("OCSF endpoint returned %d", resp.status)
        except Exception as exc:
            logger.error("Failed to send to OCSF endpoint: %s", exc)
            raise

    def _classify_event(self, event: AIBaseEvent) -> str | None:
        """Classify an OCSF event for compliance mapping."""
        class_uid = event.class_uid
        mapping = {
            7001: "model_inference",
            7002: "agent_activity",
            7003: "tool_execution",
            7004: "data_retrieval",
            7005: "security_finding",
            7006: "supply_chain",
            7007: "governance",
            7008: "identity",
        }
        return mapping.get(class_uid)

    @property
    def event_count(self) -> int:
        return self._event_count

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True
